spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSystemFunctionPushDownDQL.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.extensions;
+
+import static org.apache.iceberg.spark.SystemFunPushDownHelper.createPartitionedTable;
+import static org.apache.iceberg.spark.SystemFunPushDownHelper.days;
+import static org.apache.iceberg.spark.SystemFunPushDownHelper.hours;
+import static org.apache.iceberg.spark.SystemFunPushDownHelper.months;
+import static org.apache.iceberg.spark.SystemFunPushDownHelper.years;
+
+import java.util.List;
+import java.util.Map;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.spark.SparkSQLProperties;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.assertj.core.api.Assertions;
+import org.junit.After;
+import org.junit.Assume;
+import org.junit.Before;
+import org.junit.Test;
+
+public class TestSystemFunctionPushDownDQL extends SparkExtensionsTestBase {
+  public TestSystemFunctionPushDownDQL(
+      String catalogName, String implementation, Map<String, String> config) {
+    super(catalogName, implementation, config);
+  }
+
+  @Before
+  public void useCatalog() {
+    sql("USE %s", catalogName);
+  }
+
+  @After
+  public void removeTables() {
+    sql("DROP TABLE IF EXISTS %s PURGE", tableName);
+  }
+
+  @Test
+  public void testYearsFunction() {
+    Assume.assumeTrue(!catalogName.equals("spark_catalog"));
+
+    createPartitionedTable(spark, tableName, "years(ts)");
+
+    String date = "2017-11-22";
+    String query =
+        sqlFormat(
+            "SELECT * FROM %s WHERE system.years(ts) = system.years(date('%s')) ORDER BY id",
```

Comment:
I think the problem this is trying to solve is that you probably get unexpected results when the expression is something like `system.years(ts) = 2023`. I think this is okay and a good thing to support in general. The problem is that it still isn't obvious that the expression is misleading.

I think the right way to fix this is to allow pushing down other functions that are compatible and have the expected behavior, like Spark's built-in `year` method. That way, rather than using `system.years(ts) = 2023`, the caller could use the more-familiar `year(ts) = 2023` and it would work.

Is it possible to push down the built-in `year` method? @RussellSpitzer or @huaxingao, do you know?

We'd need to find out what gets passed as its canonical name and whether Spark will push it down.

I think we would also want to do the same thing for `month` (when combined with `year`?), `to_date`, and other conversions including `cast` if possible.

Reply:
> I think the right way to fix this is to allow pushing down other functions that are compatible and have the expected behavior, like Spark's built-in year method. That way, rather than using system.years(ts) = 2023, the caller could use the more-familiar year(ts) = 2023 and it would work.

This seems more convenient for users. There are many [time-related functions](https://github.com/apache/spark/blob/5f73b405b251f8cb39be80377e8cb2d2f960d490/sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala#L249) that can be pushed down. 
 It seems expression like following: 
```sql
year(date) = 2023
year(date) = 2023 AND month(date) = 6
year(date) = 2023 AND month(date) = 6 AND dayofmonth(date) = 5
```

Reply:
Yeah, looks like adding support for `org.apache.spark.sql.connector.expressions.Extract` and `org.apache.spark.sql.connector.expressions.Cast` is exactly what we want to do.

For now, can you remove the new rule and related codepaths? Let's just get the Iceberg system functions working with static values. You can replace the `system.years(date('%s'))` expression here with `%d` and call the transform directly to create the correct SQL value. That removes the need for the extra rule and we can add it in the next PR.

That will unblock getting this PR in. And after this PR could you add support for some of the `Extract` and `Cast` cases?

Reply:
The rule has been removed. And also removed this DQL test because we cannot push down the system function without the rule. The ut will be added with the rule in a separate rule.

>  And after this PR could you add support for some of the Extract and Cast cases?

Sure, will do.